cw_solana/lib/solana_client.dart

```diff
@@ -0,0 +1,391 @@
+    required String destinationAddress,
+    required Ed25519HDKeyPair ownerKeypair,
+    required RecentBlockhash recentBlockhash,
+    required Commitment commitment,
+  }) async {
+    // Convert SOL to lamport
+    int lamports = (inputAmount * lamportsPerSol).toInt();
+
+    final instructions = [
+      SystemInstruction.transfer(
+        fundingAccount: ownerKeypair.publicKey,
+        recipientAccount: Ed25519HDPublicKey.fromBase58(destinationAddress),
+        lamports: lamports,
+      ),
+    ];
+
+    final message = Message(instructions: instructions);
+    final signers = [ownerKeypair];
+
+    final signedTx = await _signTransactionInternal(
+      message: message,
+      signers: signers,
+      commitment: commitment,
+      recentBlockhash: recentBlockhash,
+    );
+
+    final compile = message.compile(
+      recentBlockhash: recentBlockhash.blockhash,
+      feePayer: signers.first.publicKey,
+    );
+
+    final base64Message = base64Encode(compile.toByteArray().toList());
+
+    final fee = await getGasForMessage(base64Message);
+
+    sendTx() async => await sendTransaction(
+          signedTransaction: signedTx,
+          commitment: commitment,
+        );
+
+    final pendingTransaction = PendingSolanaTransaction(
+      amount: inputAmount,
+      signedTransaction: signedTx,
+      destinationAddress: destinationAddress,
+      sendTransaction: sendTx,
+      fee: fee,
+    );
+
+    return pendingTransaction;
+  }
+
+  Future<PendingSolanaTransaction> _signSPLTokenTransaction({
+    required String tokenTitle,
+    required int tokenDecimals,
+    required String tokenMint,
+    required double inputAmount,
+    required String destinationAddress,
+    required Ed25519HDKeyPair ownerKeypair,
+    required RecentBlockhash recentBlockhash,
+    required Commitment commitment,
+  }) async {
+    final destinationOwner = Ed25519HDPublicKey.fromBase58(destinationAddress);
+    final mint = Ed25519HDPublicKey.fromBase58(tokenMint);
+
+    final associatedRecipientAccount = await _client!.getAssociatedTokenAccount(
+      mint: mint,
+      owner: destinationOwner,
+      commitment: commitment,
+    );
+    final associatedSenderAccount = await _client!.getAssociatedTokenAccount(
+      owner: ownerKeypair.publicKey,
+      mint: mint,
+      commitment: commitment,
+    );
+
+    // Throw an appropriate exception if the sender has no associated
+    // token account
+    if (associatedSenderAccount == null) {
+      throw NoAssociatedTokenAccountException(ownerKeypair.address, mint.toBase58());
+    }
+
+    // Also throw an adequate exception if the recipient has no associated
+    // token account
+    if (associatedRecipientAccount == null) {
+      throw NoAssociatedTokenAccountException(
+        destinationOwner.toBase58(),
+        mint.toBase58(),
+      );
+    }
+
+    // Input by the user
+    int userAmount = inputAmount.toInt();
+
+    int amount = int.parse('$userAmount${'0' * tokenDecimals}');
```

Comment:
Yes, this would cause some values to be lost in conversion.

Multiplying the inputAmount by 10 by tokenDecimals won't work for this case, it won't give the lamport equivalent value for this.

However, I can modify the conversion using pow instead to make it cleaner and avoid losing values in double->int conversion.

Something like this
```
Given: userAmount = 2.5 (intending to represent 2.5 units of a token), tokenDecimals = 9.

Direct multiplication: 2.5 * 10 * 9 = 225 (Won't work for this context and won't represent the conversion to the smallest token unit).

Using pow: 2.5 * pow(10, 9) = 2,500,000,000 (which correctly represents 2.5 units of the token in a unit that has 9 decimal places, such as lamports for SOL).
```

Reply:
yup, I meant pow, sorry for that confusion